File: service/src/main/java/no/nav/melosys/service/brev/bestilling/TilBrevAdresseService.java
package no.nav.melosys.service.brev.bestilling;

import no.nav.melosys.domain.Behandling;
import no.nav.melosys.domain.Kontaktopplysning;
import no.nav.melosys.domain.brev.Mottaker;
import no.nav.melosys.domain.brev.Postadresse;
import no.nav.melosys.domain.dokument.organisasjon.OrganisasjonDokument;
import no.nav.melosys.domain.person.Persondata;
import no.nav.melosys.exception.FunksjonellException;
import no.nav.melosys.integrasjon.dokgen.DokgenAdresseMapper;
import no.nav.melosys.integrasjon.ereg.EregFasade;
import no.nav.melosys.service.aktoer.KontaktopplysningService;
import no.nav.melosys.service.brev.brevmalliste.BrevAdresse;
import no.nav.melosys.service.persondata.PersondataFasade;
import org.springframework.stereotype.Component;

import static no.nav.melosys.integrasjon.dokgen.DokgenAdresseMapper.*;

@Component
public class TilBrevAdresseService {
    private final PersondataFasade persondataFasade;
    private final KontaktopplysningService kontaktopplysningService;
    private final EregFasade eregFasade;


    public TilBrevAdresseService(PersondataFasade persondataFasade, KontaktopplysningService kontaktopplysningService, EregFasade eregFasade) {
        this.persondataFasade = persondataFasade;
        this.kontaktopplysningService = kontaktopplysningService;
        this.eregFasade = eregFasade;
    }

    public BrevAdresse tilBrevAdresse(Mottaker mottaker, Behandling behandling) {
        Persondata persondata = null;
        Kontaktopplysning kontaktopplysning = null;
        OrganisasjonDokument orgDokument = null;

        switch (mottaker.getRolle()) {
            case BRUKER -> persondata = persondataFasade.hentPerson(behandling.getFagsak().hentBrukersAktørID());
            case FULLMEKTIG -> {
                if (mottaker.getPersonIdent() != null) {
                    persondata = persondataFasade.hentPerson(mottaker.getPersonIdent());
                } else {
                    kontaktopplysning = hentKontaktopplysninger(behandling, mottaker);
                    orgDokument = hentOrganisasjonsDokument(kontaktopplysning, mottaker.getOrgnr());
                }
            }
            case VIRKSOMHET, ARBEIDSGIVER -> {
                kontaktopplysning = hentKontaktopplysninger(behandling, mottaker);
                orgDokument = hentOrganisasjonsDokument(kontaktopplysning, mottaker.getOrgnr());
            }
            default -> throw new FunksjonellException("Mottakersrolle støttes ikke: " + mottaker.getRolle());
        }

        if (orgDokument == null && persondata == null) {
            throw new FunksjonellException("Orgdata eller persondata forventes for å sende brev.");
        }

        return new BrevAdresse.Builder()
            .medMottakerNavn(mapNavn(orgDokument, persondata))
            .medOrgnr(orgDokument != null ? orgDokument.getOrgnummer() : null)
            .medAdresselinjer(mapAdresselinjer(orgDokument, null, kontaktopplysning, persondata))
            .medPostnr(mapPostnr(orgDokument, persondata))
            .medPoststed(orgDokument != null ? DokgenAdresseMapper.mapPoststed(orgDokument) : mapPoststed(persondata))
            .medRegion(mapRegionForAdresse(orgDokument, persondata))
            .medLand(mapLandForAdresse(orgDokument, persondata))
            .build();
    }

    public BrevAdresse tilBrevAdresse(String personIdent, String organisasjonsnummer) {
        Persondata persondata = null;
        OrganisasjonDokument orgDokument = null;

        if (personIdent != null) {
            persondata = persondataFasade.hentPerson(personIdent);
        } else if (organisasjonsnummer != null) {
            orgDokument = hentOrganisasjonsDokument(null, organisasjonsnummer);
        }

        if (orgDokument == null && persondata == null) {
            throw new FunksjonellException("Orgdata eller persondata forventes for å finne adresse.");
        }


        return new BrevAdresse.Builder()
            .medMottakerNavn(mapNavn(orgDokument, persondata))
            .medOrgnr(orgDokument != null ? orgDokument.getOrgnummer() : null)
            .medAdresselinjer(mapAdresselinjer(orgDokument, null, null, persondata))
            .medPostnr(mapPostnr(orgDokument, persondata))
            .medPoststed(orgDokument != null ? DokgenAdresseMapper.mapPoststed(orgDokument) : mapPoststed(persondata))
            .medRegion(mapRegionForAdresse(orgDokument, persondata))
            .medLand(mapLandForAdresse(orgDokument, persondata))
            .build();
    }

    private Kontaktopplysning hentKontaktopplysninger(Behandling behandling, Mottaker mottaker) {
        return kontaktopplysningService.hentKontaktopplysning(behandling.getFagsak().getSaksnummer(), mottaker.getOrgnr()).orElse(null);
    }

    private OrganisasjonDokument hentOrganisasjonsDokument(Kontaktopplysning kontaktopplysning, String orgnr) {
        String mottakerOrgnr = kontaktopplysning != null && kontaktopplysning.getKontaktOrgnr() != null
            ? kontaktopplysning.getKontaktOrgnr() : orgnr;
        return (OrganisasjonDokument) eregFasade.hentOrganisasjon(mottakerOrgnr).getDokument();
    }

    private String mapPoststed(Persondata persondata) {
        Postadresse postadresse = persondata.hentGjeldendePostadresse();
        if (postadresse == null) {
            return null;
        }
        return postadresse.poststed();
    }
}

File: service/src/main/kotlin/no/nav/melosys/service/brev/TilBrevAdresseService.kt
package no.nav.melosys.service.brev

import no.nav.melosys.domain.Behandling
import no.nav.melosys.domain.Kontaktopplysning
import no.nav.melosys.domain.brev.Mottaker
import no.nav.melosys.domain.dokument.organisasjon.OrganisasjonDokument
import no.nav.melosys.domain.kodeverk.Mottakerroller
import no.nav.melosys.domain.person.Persondata
import no.nav.melosys.exception.FunksjonellException
import no.nav.melosys.integrasjon.dokgen.DokgenAdresseMapper
import no.nav.melosys.integrasjon.ereg.EregFasade
import no.nav.melosys.service.aktoer.KontaktopplysningService
import no.nav.melosys.service.persondata.PersondataFasade
import org.apache.commons.lang3.StringUtils
import org.springframework.stereotype.Component


@Component
class TilBrevAdresseService(
    private val persondataFasade: PersondataFasade,
    private val kontaktopplysningService: KontaktopplysningService,
    private val eregFasade: EregFasade
) {
    fun tilBrevAdresse(mottaker: Mottaker, behandling: Behandling): BrevAdresse {
        var persondata: Persondata? = null
        var kontaktopplysning: Kontaktopplysning? = null
        var orgDokument: OrganisasjonDokument? = null

        when (mottaker.rolle) {
            Mottakerroller.BRUKER -> persondata = persondataFasade.hentPerson(behandling.fagsak.hentBrukersAktørID())
            Mottakerroller.FULLMEKTIG -> {
                if (mottaker.personIdent != null) {
                    persondata = persondataFasade.hentPerson(mottaker.personIdent)
                } else {
                    kontaktopplysning = hentKontaktopplysninger(behandling, mottaker)
                    orgDokument = hentOrganisasjonsDokument(kontaktopplysning, mottaker.orgnr)
                }
            }

            Mottakerroller.VIRKSOMHET, Mottakerroller.ARBEIDSGIVER -> {
                kontaktopplysning = hentKontaktopplysninger(behandling, mottaker)
                orgDokument = hentOrganisasjonsDokument(kontaktopplysning, mottaker.orgnr)
            }

            else -> throw FunksjonellException("Mottakersrolle støttes ikke: " + mottaker.rolle)
        }

        if (orgDokument == null && persondata == null) {
            throw FunksjonellException("Orgdata eller persondata forventes for å sende brev.")
        }

        return BrevAdresse(
            mottakerNavn = DokgenAdresseMapper.mapNavn(orgDokument, persondata),
            orgnr = orgDokument?.orgnummer,
            adresselinjer = DokgenAdresseMapper.mapAdresselinjer(orgDokument, null, kontaktopplysning, persondata)
                ?.filterNot { StringUtils.isEmpty(it) },
            postnr = DokgenAdresseMapper.mapPostnr(orgDokument, persondata),
            poststed = if (orgDokument != null) DokgenAdresseMapper.mapPoststed(orgDokument) else mapPoststed(persondata!!),
            region = DokgenAdresseMapper.mapRegionForAdresse(orgDokument, persondata),
            land = DokgenAdresseMapper.mapLandForAdresse(orgDokument, persondata),
        )
    }

    fun tilBrevAdresse(personIdent: String?, organisasjonsnummer: String?): BrevAdresse {
        var persondata: Persondata? = null
        var orgDokument: OrganisasjonDokument? = null

        if (personIdent != null) {
            persondata = persondataFasade.hentPerson(personIdent)
                ?: throw FunksjonellException("Finner ikke persondata for personIdent.")
        } else if (organisasjonsnummer != null) {
            orgDokument = hentOrganisasjonsDokument(null, organisasjonsnummer)
        } else {
            throw FunksjonellException("Kan ikke finne adresse uten personIdent og organisasjonsnummer")
        }

        return BrevAdresse(
            mottakerNavn = DokgenAdresseMapper.mapNavn(orgDokument, persondata),
            orgnr = orgDokument?.orgnummer,
            adresselinjer = DokgenAdresseMapper.mapAdresselinjer(orgDokument, null, null, persondata)?.filterNot { StringUtils.isEmpty(it) },
            postnr = DokgenAdresseMapper.mapPostnr(orgDokument, persondata),
            poststed = if (orgDokument != null) DokgenAdresseMapper.mapPoststed(orgDokument) else mapPoststed(persondata!!),
            region = DokgenAdresseMapper.mapRegionForAdresse(orgDokument, persondata),
            land = DokgenAdresseMapper.mapLandForAdresse(orgDokument, persondata)
        )
    }

    private fun hentKontaktopplysninger(behandling: Behandling, mottaker: Mottaker): Kontaktopplysning? {
        return kontaktopplysningService.hentKontaktopplysning(behandling.fagsak.saksnummer, mottaker.orgnr).orElse(null)
    }

    private fun hentOrganisasjonsDokument(kontaktopplysning: Kontaktopplysning?, orgnr: String): OrganisasjonDokument {
        val mottakerOrgnr = kontaktopplysning?.kontaktOrgnr ?: orgnr
        return eregFasade.hentOrganisasjon(mottakerOrgnr).dokument as OrganisasjonDokument
    }

    private fun mapPoststed(persondata: Persondata): String? {
        return persondata.hentGjeldendePostadresse()?.poststed
    }
}
